decompile.py

```diff
@@ -40,33 +41,30 @@ def from_raw(cls, data, frontend: Frontend = BinaryninjaFrontend) -> Decompiler:
         """Create a decompiler instance from existing frontend instance (e.g. a binaryninja view)."""
         return cls(frontend.from_raw(data))
 
-    def decompile(self, function: str, task_options: Optional[Options] = None) -> DecompilerTask:
-        """Decompile the target function."""
-        # Sanity check to ensure task_options is populated
+    def decompile(self, function_ids: Collection[object] | None = None, task_options: Options | None = None) -> Result:
+        if function_ids is None:  # decompile all functions when none are specified
+            function_ids = self._frontend.get_all_function_names()
         if task_options is None:
             task_options = Decompiler.create_options()
-        # Start decompiling
-        pipeline = DecompilerPipeline.from_strings(task_options.getlist("pipeline.cfg_stages"), task_options.getlist("pipeline.ast_stages"))
-        task = self._frontend.create_task(function, task_options)
-        pipeline.run(task)
-        task.code = self._backend.generate([task])
-        return task
-
-    def decompile_all(self, task_options: Optional[Options] = None) -> str:
-        """Decompile all functions in the binary"""
-        tasks = list()
-        # Sanity check to ensure task_options is populated
-        if task_options is None:
-            task_options = Decompiler.create_options()
-        # Start decompiling
+
         pipeline = DecompilerPipeline.from_strings(task_options.getlist("pipeline.cfg_stages"), task_options.getlist("pipeline.ast_stages"))
-        functions = self._frontend.get_all_function_names()
-        for function in functions:
-            task = self._frontend.create_task(function, task_options)
-            pipeline.run(task)
+
+        tasks = []
+        for function_id in function_ids:
+            task = DecompilerTask(str(function_id), function_id, task_options)
             tasks.append(task)
+
+            self._frontend.lift(task)
```

Comment:
Why the split between creating the decompiler-task and lifting it?

Reply:
There is not much of a motivation behind creating the `DecompilerTask` outside the lift method.
The decision arose from the fact, that the creation of the task object had to move to the front of the lifting procedure, as any exception before the existence of the decompiler task object meant we had to handle the error different, than after the creation of the task object.
I simply decided to put the creation of the task object outside the lift method instead of at the beginning of the lift function, as I though it made more sense, that the lift method simply deals with lifting data into the task.